core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -709,9 +713,10 @@ class DAGScheduler(
       // cancelling the stages because if the DAG scheduler is stopped, the entire application
       // is in the process of getting stopped.
       val stageFailedMessage = "Stage cancelled because SparkContext was shut down"
-      runningStages.foreach { stage =>
-        stage.latestInfo.stageFailed(stageFailedMessage)
-        listenerBus.post(SparkListenerStageCompleted(stage.latestInfo))
+      // The `toArray` here is necessary so that we don't iterate over `runningStages` while
+      // mutating it.
+      runningStages.toArray.foreach { stage =>
```

Comment:
You could avoid the comment by doing:

    while (!runningStages.isEmpty) markStageAsFinished(runningStages.last)

But either is fine.